persistent/Database/Persist/Quasi/Internal/ModelParser.hs

```diff
@@ -258,37 +257,118 @@ data SourceLoc = SourceLoc
     }
     deriving (Show, Lift)
 
--- @since 2.16.0.0
-data Token
-    = Quotation Text
-    | Equality Text Text
+-- | An attribute of an entity field definition.
+--
+-- @since 2.17.1.0
+data Attribute
+    = Assignment Text Text
     | Parenthetical Text
-    | BlockKey Text
     | PText Text
+    | -- | Quoted field attributes are deprecated since 2.17.1.0.
+      Quotation Text
+    deriving (Eq, Ord, Show)
+
+-- | An argument to an entity field directive.
+--
+-- @since 2.17.1.0
+data DirectiveArgument
+    = -- This is too unstructured. We should rework directive parsing and make this smarter.
+      DText Text
+    | -- | Quoted directive arguments are deprecated since 2.17.1.0.
+      DQuotation Text
     deriving (Eq, Ord, Show)
 
+-- | The name of an entity block or extra block.
+--
+-- @since 2.17.1.0
+newtype BlockKey = BlockKey Text
+    deriving (Show)
+
+-- | A parsed comment or doc comment.
+--
 -- @since 2.16.0.0
 data CommentToken
     = DocComment Text
     | Comment Text
     deriving (Eq, Ord, Show)
 
--- | Converts a token into a Text representation for second-stage parsing or presentation to the user
+-- | Converts an attribute into a Text representation for second-stage parsing or
+-- presentation to the user
 --
 -- @since 2.16.0.0
-tokenContent :: Token -> Text
-tokenContent = \case
-    Quotation s -> s
-    Equality l r -> mconcat [l, "=", r]
+attributeContent :: Attribute -> Text
+attributeContent = \case
+    Assignment l r -> mconcat [l, "=", r]
     Parenthetical s -> s
     PText s -> s
-    BlockKey s -> s
+    Quotation s -> s
+
+-- | Converts a directive into a Text representation for second-stage parsing or
+-- presentation to the user
+--
+-- @since 2.17.1.0
+directiveContent :: Directive -> [Text]
+directiveContent d = directiveArgumentContent <$> directiveArguments d
+
+entityFieldContent :: EntityField -> [Text]
+entityFieldContent f =
+    [ fieldNameAndStrictnessAsText f
+    , (typeExprContent . entityFieldType) f
+    ]
+        ++ fmap attributeContent (entityFieldAttributes f)
+
+blockKeyContent :: BlockKey -> Text
+blockKeyContent (BlockKey t) = t
+
+-- | Generates the field name of an EntityField, accompanied by
+-- its strictness sigil, if one is present.
+-- This is only needed temporarily, and can eventually be refactored away.
+--
+-- @since 2.17.1.0
+fieldNameAndStrictnessAsText :: EntityField -> Text
+fieldNameAndStrictnessAsText f =
+    let
+        s = case entityFieldStrictness f of
+            Just Strict -> "!"
+            Just Lazy -> "~"
+            Nothing -> ""
+        (FieldName fn) = entityFieldName f
+     in
+        s <> fn
 
 commentContent :: CommentToken -> Text
 commentContent = \case
     Comment s -> s
     DocComment s -> s
 
+quotedAttributeErrorMessage :: String
```

Comment:
We define a custom error message mostly so that we can fish it out again in `tryOrReport`, because it's otherwise not as easy as you'd think to determine when the quoted-attribute-forbidding parser has failed because of a quoted attribute as opposed to some other reason.

I don't love this but it works fine for now; a better solution would probably add 100+ LOC to this already-big PR.